codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -114,16 +169,136 @@ private void generateOperationSerializer(
             GenerationContext context,
             OperationShape operation
     ) {
+        generateOperationSerializerMiddleware(context, operation);
         generateOperationHttpBindingSerializer(context, operation);
+        generateOperationDocumentSerializer(context, operation);
+        addOperationDocumentShapeBinders(context, operation);
+    }
+
+    protected abstract void generateOperationDocumentSerializer(GenerationContext context, OperationShape operation);
+
+    /**
+     * Adds the top-level shapes from the operation that bind to the body document that require serializer functions.
+     *
+     * @param context   the generator context
+     * @param operation the operation to add document binders from
+     */
+    private void addOperationDocumentShapeBinders(GenerationContext context, OperationShape operation) {
+        Model model = context.getModel();
+
+        /*
+         * Walk and members shapes to the list that will require serializer functions
+         */
+        List<HttpBinding> bindings = new ArrayList<>(model.getKnowledge(HttpBindingIndex.class)
+                .getRequestBindings(operation).values());
+
+        for (HttpBinding binding : bindings) {
+            Shape targetShape = model.expectShape(binding.getMember().getTarget());
+            if (isShapeTypeDocumentSerializerRequired(targetShape.getType())
+                    && (binding.getLocation() == HttpBinding.Location.DOCUMENT
+                    || binding.getLocation() == HttpBinding.Location.PAYLOAD)) {
+                serializeDocumentBindingShapes.add(targetShape);
+            }
+        }
     }
 
+    private void generateOperationSerializerMiddleware(GenerationContext context, OperationShape operation) {
+        GoStackStepMiddlewareGenerator middleware = GoStackStepMiddlewareGenerator.createSerializeStepMiddleware(
+                ProtocolGenerator.getSerializeMiddlewareName(operation.getId(), getProtocolName()
+                ));
+
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Model model = context.getModel();
+
+        Shape inputShape = model.expectShape(operation.getInput()
+                .orElseThrow(() -> new CodegenException("expect input shape for operation: " + operation.getId())));
+
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+
+        ApplicationProtocol applicationProtocol = getApplicationProtocol();
+        Symbol requestType = applicationProtocol.getRequestType();
+
+        // TODO: Smithy http binding code generator should not know AWS types, composition would help solve this
+        middleware.writeMiddleware(context.getWriter(), (generator, writer) -> {
+            writer.addUseImports(requestType);
```

Comment:
These will get added automagically when you use the symbol with the T or P formatters